File: code/Kotlin/libs/annotation_processing/apt/compiler/src/main/java/inject/processor/creator/CreatorProcessor.java
package inject.processor.creator;

import com.google.auto.service.AutoService;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeSpec;

import java.lang.annotation.Annotation;
import java.util.ArrayList;
import java.util.List;

import javax.annotation.processing.Processor;
import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.type.TypeMirror;

import inject.annotation.creator.Creator;
import inject.annotation.creator.Extra;
import inject.classname.AndroidName;
import inject.classname.VectorName;
import inject.kotlin.KotlinClassName;
import inject.processor.BaseProcessor;

import static javax.lang.model.element.Modifier.FINAL;
import static javax.lang.model.element.Modifier.PRIVATE;
import static javax.lang.model.element.Modifier.PUBLIC;

/**
 * @author yuansui
 * @since 2017/8/14
 */
@AutoService(Processor.class)
//@Incremental(IncrementalType.AGGREGATING)
public class CreatorProcessor extends BaseProcessor {

    @Override
    public Class<? extends Annotation> getAnnotationClass() {
        return Creator.class;
    }

    @Override
    public TypeSpec createTypeSpec(Element annotatedElement) {
        CommonMethod commonMethod = new CommonMethod(this, annotatedElement);

        TypeMirror annotatedMirror = annotatedElement.asType();

        final String fileName = String.format("%sCreator", annotatedElement.getSimpleName());
        TypeSpec.Builder builder = TypeSpec.classBuilder(fileName)
                .addModifiers(PUBLIC, FINAL);

        // 添加成员变量
        List<Element> all = new ArrayList<>();
        getAnnotatedFields(annotatedElement, all);
        for (Element e : all) {
            String paramName = getParamName(e);
            builder.addField(getTypeNameBox(e), paramName, PRIVATE);
        }

        builder.addMethod(commonMethod.constructor());
        builder.addMethod(commonMethod.create(fileName));
        builder.addMethods(commonMethod.optionalSetters(fileName));

        if (isSubtype(annotatedMirror, getElementTypeMirror(AndroidName.ACTIVITY))) {
            // activity
            builder.addMethod(commonMethod.newIntent());

            ActivityMethod method = new ActivityMethod(this, annotatedElement);
            Creator c = annotatedElement.getAnnotation(Creator.class);
            builder.addMethod(method.start(c.withTransition()));
            if (c.forResult()) {
                builder.addMethod(method.startForResult());
            }
        } else if (isSubtype(annotatedMirror, getElementTypeMirror(AndroidName.FRAGMENT))) {
            // fragment
            FragmentMethod method = new FragmentMethod(this, annotatedElement);
            builder.addMethod(method.get());
        } else if (isSubtype(annotatedMirror, getElementTypeMirror(AndroidName.SERVICE))) {
            // Service
            if (isSubtype(annotatedMirror, getElementTypeMirror(VectorName.SERVICE))) {
                builder.addMethod(commonMethod.newIntent());

                ServiceMethod method = new ServiceMethod(this, annotatedElement);
                builder.addMethod(method.start());
                builder.addMethod(method.startForeground());
                builder.addMethod(method.stop());
            } else if (isSubtype(annotatedMirror, getElementTypeMirror(VectorName.JOB_SERVICE))) {
                JobServiceMethod method = new JobServiceMethod(this, annotatedElement);
                builder.addMethod(method.start());
                builder.addMethod(method.start2());
            } else if (isSubtype(annotatedMirror, getElementTypeMirror(VectorName.JOB_INTENT_SERVICE))) {
                builder.addMethod(commonMethod.newIntent());

                JobIntentServiceMethod method = new JobIntentServiceMethod(this, annotatedElement);
                builder.addMethod(method.start());
            }
        }

        builder.addMethod(commonMethod.inject());

        return builder.build();
    }

    private void getAnnotatedFields(Element annotatedElement, List<Element> all) {
        for (Element e : annotatedElement.getEnclosedElements()) {
            Extra a = e.getAnnotation(Extra.class);
            if (a != null) {
                all.add(e);
            }
        }

        List<? extends TypeMirror> superTypes = getTypeUtils().directSupertypes(annotatedElement.asType());
        TypeMirror superClassType = !superTypes.isEmpty() ? superTypes.get(0) : null;
        Element superClass = superClassType == null ? null : getTypeUtils().asElement(superClassType);
        if (superClass != null && superClass.getKind() == ElementKind.CLASS) {
            getAnnotatedFields(superClass, all);
        }
    }

    TypeMirror getElementTypeMirror(CharSequence name) {
        return getElementUtils().getTypeElement(name).asType();
    }

    String getParamName(Element e) {
        return getParamName(e, null);
    }

    String getIntentParamsText(List<Element> all) {
        StringBuilder sb = new StringBuilder();
        sb.append("$T intent = newIntent(context");
        for (Element e : all) {
            String paramName = getParamName(e);
            sb.append(",").append(paramName);
        }
        sb.append(")");
        return sb.toString();
    }

    void addEson(MethodSpec.Builder builder) {
        builder.addStatement("$T eson = $T.getDefault()", KotlinClassName.ESON, KotlinClassName.ESON);
    }
}

File: code/Kotlin/libs/annotation_processing/kapt/compiler/src/main/java/inject/compiler/processor/CreatorProcessor.kt
package inject.compiler.processor

import com.google.auto.service.AutoService
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.KModifier
import com.squareup.kotlinpoet.PropertySpec
import com.squareup.kotlinpoet.TypeSpec
import inject.annotation.creator.Creator
import inject.annotation.creator.Extra
import inject.compiler.classname.AndroidName
import inject.compiler.method.creator.ActivityMethod
import inject.compiler.method.creator.CommonMethod
import inject.compiler.method.creator.FragmentMethod
import inject.compiler.method.creator.ServiceMethod
import inject.compiler.ext.asTypeName
import inject.compiler.ext.getParamName
import javax.annotation.processing.Processor
import javax.lang.model.element.Element
import javax.lang.model.element.ElementKind
import javax.lang.model.type.TypeMirror

/**
 * @author yuansui
 * @since 2017/8/14
 */
@AutoService(Processor::class)
class CreatorProcessor : BaseProcessor() {

    override fun getAnnotationClass(): Class<out Annotation> {
        return Creator::class.java
    }

    override fun createTypeSpec(annotatedElement: Element): TypeSpec {
        val commonMethod = CommonMethod(this, annotatedElement)

        val annotatedMirror = annotatedElement.asType()

        val fileName = String.format("%sCreator", annotatedElement.simpleName)
        val builder = TypeSpec.Companion.classBuilder(fileName)
            .primaryConstructor(
                FunSpec.Companion.constructorBuilder()
                    .addModifiers(KModifier.PRIVATE)
                    .build()
            )

        // 添加成员变量
        val all = mutableListOf<Element>()
        getAnnotatedFields(annotatedElement, all)
        for (e in all) {
            val paramName = e.getParamName()
            builder.addProperty(
                PropertySpec.Companion.builder(paramName, e.asTypeName().copy(nullable = true))
                    .addModifiers(KModifier.PRIVATE)
                    .mutable()
                    .initializer("null")
                    .build()
            )
        }

        builder.addType(commonMethod.create(fileName))
        builder.addFunctions(commonMethod.optionalSetters(fileName))

        when {
            isSubtype(annotatedMirror, getElementTypeMirror(AndroidName.ACTIVITY)) -> {
                // activity
                builder.addFunction(commonMethod.newIntent())

                val method = ActivityMethod(this, annotatedElement)
                val c = annotatedElement.getAnnotation<Creator>(Creator::class.java)
                builder.addFunction(method.start(c.withTransition))
                if (c.forResult) {
                    builder.addFunction(method.startForResult())
                }
            }

            isSubtype(annotatedMirror, getElementTypeMirror(AndroidName.FRAGMENT)) -> {
                // fragment
                val method = FragmentMethod(this, annotatedElement)
                builder.addFunction(method.get())
            }

            isSubtype(annotatedMirror, getElementTypeMirror(AndroidName.SERVICE)) -> {
                // Service
                builder.addFunction(commonMethod.newIntent())

                val method = ServiceMethod(this, annotatedElement)
                builder.addFunction(method.start())
                builder.addFunction(method.startForeground())
//                builder.addType(method.stop())
            }
        }

        return builder.build()
    }

    private fun getAnnotatedFields(annotatedElement: Element, all: MutableList<Element>) {
        for (e in annotatedElement.enclosedElements) {
            val a = e.getAnnotation<Extra>(Extra::class.java)
            if (a != null) {
                all.add(e)
            }
        }

        val superTypes: MutableList<out TypeMirror?> = getTypeUtils().directSupertypes(annotatedElement.asType())
        val superClassType: TypeMirror? = if (!superTypes.isEmpty()) superTypes[0] else null
        val superClass = if (superClassType == null) null else getTypeUtils().asElement(superClassType)
        if (superClass != null && superClass.kind == ElementKind.CLASS) {
            getAnnotatedFields(superClass, all)
        }
    }

    fun getElementTypeMirror(name: CharSequence): TypeMirror {
        return getElementUtils().getTypeElement(name).asType()
    }

    fun getIntentParamsText(all: List<Element>): String {
        val sb = StringBuilder()
        sb.append("val intent = newIntent(context")
        for (e in all) {
            val paramName = e.getParamName()
            sb.append(",").append(paramName)
        }
        sb.append(")")
        return sb.toString()
    }
}